File: python/leetcode/095.py
class Solution(object):
    def generateTrees(self, n):
        if n == 0:
            return []
        return self.dfs(1, n)

    def dfs(self, start, end):
        if start > end:
            return [None]
        res = []
        for idx in range(start, end + 1):
            leftnodes = self.dfs(start, idx - 1)
            rightnodes = self.dfs(idx + 1, end)
            for left in leftnodes:
                for right in rightnodes:
                    root = TreeNode(val)
                    root.left = left
                    root.right = right
                    res.append(root)
        return res

File: cpp/leetcode/095.cpp
class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        if (n == 0) {
            return {};
        }
        return dfs(1, n);
    }

    vector<TreeNode *> dfs(int start, int end) {
        vector<TreeNode *> res;
        if (start > end) {
            res.push_back(NULL);
            return res;
        }
        for (int i = start; i <= end; i++) {
            for (auto left : dfs(start, i - 1)) {
                for (auto right : dfs(i + 1, end)) {
                    TreeNode *root = new TreeNode(i);
                    root->left = left;
                    root->right = right;
                    res.push_back(root);
                }
            }
        }
        return res;
    }
};